iothub/device/devdoc/API Design.md

```diff
@@ -0,0 +1,255 @@
+
+### Properties
+
+```csharp
+/// <summary>
+/// Update a single property.
+/// 
+/// Property under the root component:
+/// "reported": {
+///     "temperature": 21.3
+/// }
+/// 
+/// Property under a specified component:
+/// "reported": {
+///     "thermostat1": {
+///         "__t": "c",
+///         "temperature": 21.3
+///     }
+/// }
+/// </summary>
+/// <param name="propertyName">The name of the twin property to update.</param>
+/// <param name="propertyValue">The value of the twin property to update.</param>
+/// <param name="componentName">The name of the component in which the property is defined. Can be null for property defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, dynamic propertyValue, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Update a batch of properties.
+/// 
+/// Properties under the root component:
+/// "reported": {
+///     "temperature": 21.3,
+///     "humudity": 60
+/// }
+/// 
+/// Properties under a specified component:
+/// "reported": {
+///     "thermostat1": {
+///         "__t": "c",
+///         "temperature": 21.3,
+///         "humudity": 60
+///     }
+/// }
+/// </summary>
+/// <param name="properties">The twin properties and values to update.</param>
+/// <param name="componentName">The name of the component in which the properties are defined. Can be null for properties defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertiesAsync(IDictionary<string, dynamic> properties, string componentName = default, CancellationToken cancellationToken = default);
+
+/// <summary>
+/// Respond to a writable property request received from the service.
+/// 
+/// Property under the root component:
+/// "reported": {
+///     "targetTemperature": {
+///         "value": 21.3,
+///         "ac": 200,
+///         "av": 3,
+///         "ad": "complete"
+///     }
+/// }
+/// 
+/// Property under a specified component:
+/// "reported": {
+///   "thermostat1": {
+///     "__t": "c",
+///     "targetTemperature": {
+///       "value": 23,
+///       "ac": 200,
+///       "av": 3,
+///       "ad": "complete"
+///     }
+///   }
+/// }
+/// </summary>
+/// <param name="propertyName">The name of the twin property to update.</param>
+/// <param name="propertyValue">The value of the twin property to update.</param>
+/// <param name="componentName">The name of the component in which the property is defined. Can be null for property defined under the root interface.</param>
+/// <param name="cancellationToken">A cancellation token to cancel the operation.</param>
+public Task UpdatePropertyAsync(string propertyName, WritableProperty propertyValue, string componentName = default, CancellationToken cancellationToken = default);
```

Comment:
This seems to be for responding to a writeable property update?  And it's an overload of setting a "read-only" property.  Correct??

If so, I thought we had decided to give this a different name (like RespondToWriteableProperty??) since overload had pros/cons.

If we've changed there maybe good reasons but I'd like to understand more behind the rationale.

Reply:
Yes, this usually sent as a response to a "writable" property update but doesn't need to be necessarily; eg: if a device starts up and sends an initial value for a "writable property"

https://docs.microsoft.com/en-us/azure/iot-pnp/concepts-convention#writable-properties
"When a device starts up for the first time, it can send an initial value for a reported property if it doesn't receive an initial desired property from the hub. In this case, the device should set av to 1."

It will still always be for a "writable" property, so I don't mind adding that term in here.
I might have missed the name change in my notes, there wasn't a deliberate reason for me calling this `UpdatePropertyAsync`

Reply:
`RespondToWritablePropertyEvent` would be the natural flow here since we have `SubscribeToWritablePropertyEvent`

Reply:
@abhipsaMisra Great point about the RespondTo... naming. Either we use the overload or we create a specific UpdateWritableProperty method.